### PR TITLE
Focus Cluster List

### LINKED ISSUE #1121: Filter/Sort for cluster with persisting issues
**What would you like to be added**:
As an operator I want to have an option to filter and/or sort the `All Projects` view for Shoot clusters with **persisting** issues.

- An option to filter the list for issues that persist longer that `5m`, `10m`, `1h` (or even configurable timeframe) would be preferable.
- An option to sort the list for oldest cluster with issue to newest cluster with issues would be helpful (considering the Shoot status and the Shoot conditions).
- Furthermore it would be great if the update of the list can be disabled for a certain point in time.

**Why is this needed**:
As Gardener comes with self healing capabilities therefore many issues resolve on its own after some time e.g. with the next retried Shoot operation. As an operator I might want to give the system a chance to self heal before digging into the issue. 

/cc @ScheererJ, @mliepold, @dguendisch

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
Allow operators to freeze the all all projects cluster list
- Clusters no longer change position, sorting is fixed
- Items still get updates
- Gone clusters are kept as stale objects in the list
- New clusters are not added but a badge next to switch informs about new clusters

Screen Shot 2022-07-12 at 13 54 57

**Which issue(s) this PR fixes**:
Fixes #1121

**Special notes for your reviewer**:
Build currently fails due to coverage not met

**Release note**:

```feature operator
New focus mode for the `all projects` cluster list: This feature enables operators to get a static overview of clusters with issues. When the focus mode is active, the cluster list behavior changes:
- Clusters no longer change position, the sorting is fixed even if data changes
- Gone clusters are kept as stale objects in the list to keep the list steady
- New clusters are not added but a badge next to the focus switch informs about new clusters
```

### CODE REVIEW

File: frontend/tests/unit/utils.spec.js [hunk 1]
```diff
@@ -521,5 +521,95 @@ describe('utils', () => {
...
+        expect(getIssueSince({...sampleStatuses.statusConstraints})).toBe('2000-00-00T00:00:06Z')
+      })
+
+      it('should return issue since for lastError', () => {
+        expect(getIssueSince({...sampleStatuses.statusLastErrors})).toBe('2000-00-00T00:00:08Z')
+      })
+
+      it('should return issue since for allIssues', () => {
+        expect(getIssueSince({...sampleStatuses.statusAll})).toBe('2000-00-00T00:00:01Z')
+      })
```
Comment (holgerkoser) [inline frontend/tests/unit/utils.spec.js:612]:
```suggestion
      let status

      beforeEach(() => {
        status = {
          lastOperation: {
            state: 'False',
            lastUpdateTime: '2000-01-01T00:00:01Z'
          },
          conditions: [
            {
              status: 'True',
              lastTransitionTime: '2000-01-01T00:00:02Z'
            },
            {
              status: 'False',
              lastTransitionTime: '2000-01-01T00:00:03Z'
            },
            {
              status: 'False',
              lastTransitionTime: '2000-01-01T00:00:04Z'
            }
          ],
          constraints: [
            {
              status: 'True',
              lastTransitionTime: '2000-01-01T00:00:05Z'
            },
            {
              status: 'False',
              lastTransitionTime: '2000-01-01T00:00:06Z'
            },
            {
              status: 'False',
              lastTransitionTime: '2000-01-01T00:00:07Z'
            }
          ],
          lastErrors: [
            {
              lastUpdateTime: '2000-01-01T00:00:08Z'
            }
          ]
        }
      })

      it('should not fail when zero', () => {
        expect(getIssueSince()).toBeUndefined()
      })

      it('should return undefined for lastOperation == true', () => {
        status.lastOperation.state = 'True'
        expect(getIssueSince(pick(status, 'lastOperation'))).toBeUndefined()
      })

      it('should return issue since for lastOperation', () => {
        expect(getIssueSince(pick(status, 'lastOperation'))).toBe('2000-01-01T00:00:01Z')
      })

      it('should return issue since for condition', () => {
        expect(getIssueSince(pick(status, 'conditions'))).toBe('2000-01-01T00:00:03Z')
      })

      it('should return issue since for constraint', () => {
        expect(getIssueSince(pick(status, 'constraints'))).toBe('2000-01-01T00:00:06Z')
      })

      it('should return issue since for lastError', () => {
        expect(getIssueSince(pick(status, 'lastErrors'))).toBe('2000-01-01T00:00:08Z')
      })

      it('should return issue since for allIssues', () => {
        expect(getIssueSince(status)).toBe('2000-01-01T00:00:01Z')
      })
```

File: frontend/src/views/ShootList.vue [hunk 2]
```diff
@@ -16,6 +16,41 @@ SPDX-License-Identifier: Apache-2.0
...
+                  </template>
+                </v-switch>
+              </v-badge>
+            </div>
+          </template>
+          <span class="font-weight-bold">Focus Mode</span><br />
+          Cluster List sorting is freezed.<br />
+          Items in the list will still be updated.<br />
+          New items will not be added to the list.<br />
+          Removed items will be shown as stale (greyed out).
```
Comment (petersutter) [inline frontend/src/views/ShootList.vue:47]: as list items

File: frontend/src/views/ShootList.vue [hunk 3]
```diff
@@ -16,6 +16,41 @@ SPDX-License-Identifier: Apache-2.0
...
+          <template v-if="numberOfNewItemsSinceFreeze > 0">
+            <v-divider color="white"></v-divider>
+            <span class="font-weight-bold">{{numberOfNewItemsSinceFreeze}}</span> new clusters were added to the list since you enabled focus mode.<br />
+            New clusters will not be added to the list until you disable focus mode.
```
Comment (petersutter) [inline frontend/src/views/ShootList.vue:51]:
```suggestion
            New clusters with issues will not be added to the list until you disable focus mode.
```
also add to list items and show always

File: frontend/src/views/ShootList.vue [hunk 4]
```diff
@@ -16,6 +16,41 @@ SPDX-License-Identifier: Apache-2.0
           <div class="text-subtitle-1 toolbar-title--text">{{headlineSubtitle}}</div>
         </v-toolbar-title>
         <v-spacer></v-spacer>
+        <v-tooltip top>
```
Comment (petersutter) [inline frontend/src/views/ShootList.vue:19]:
```suggestion
        <v-tooltip bottom>
```

otherwise the tooltip overlaps the badge

File: frontend/src/views/ShootList.vue (hunk 2 above)
Comment (grolu) [reply frontend/src/views/ShootList.vue:47]: done

File: frontend/src/store/modules/shoots/index.js [hunk 5]
```diff
@@ -364,6 +366,35 @@ const actions = {
 
     commit('RESET_NEW_SHOOT_RESOURCE', shootResource)
     return state.newShootResource
+  },
+  setFreezeSorting ({ commit }, value) {
+    commit('SET_FREEZE_SORTING', value)
+
+    if (value) {
+      const freezedShootSkeletons = map(state.filteredShoots, shoot => {
```
Comment (petersutter) [inline frontend/src/store/modules/shoots/index.js:374]: why "only" skeletons?

File: frontend/src/store/modules/shoots/index.js [hunk 6]
```diff
@@ -364,6 +366,35 @@ const actions = {
...
+            'status.technicalID',
+            'spec.kubernetes.version',
+            'status' // need to include not only values that we want to display, also need data for sorting
+          ]),
+          stale: true // Flag them so that the UI can render them accordingly
+        }
+      })
+
+      commit('SET_FREEZED_SHOOT_SKELETONS', freezedShootSkeletons)
+    }
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/store/modules/shoots/index.js:395]: added logic to set to undefined if value is false

File: frontend/src/store/modules/shoots/getters.js [hunk 7]
```diff
@@ -257,11 +297,44 @@ export default {
...
+          itemsToSort = orderBy(itemsToSort, [item => getSortVal(rootGetters, item, sortBy), 'metadata.name'], [sortOrder, 'asc'])
         }
       }
-      return items
+
+      sortedUIDsAtFreeze = map(itemsToSort, ({ metadata }) => {
+        return metadata.uid
+      })
+      sortByArrAtFreeze = sortByArr
+      sortDescArrAtFreeze = sortDescArr
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/store/modules/shoots/getters.js:308]: I incorporated the feedback.
- The variables are now set outside of the sort function
- Copy the array values and compare the arrays instead of the pointers (to prevent this logic to break if vuetify changes the internal logic)

File: frontend/src/store/modules/shoots/getters.js [hunk 8]
```diff
@@ -206,11 +220,37 @@ export default {
     }
   },
   sortItems (state, getters, rootState, rootGetters) {
-    return (items, sortByArr, sortDescArr) => {
+    let sortedUIDsAtFreeze
+    let sortByArrAtFreeze
+    let sortDescArrAtFreeze
+
+    const sortFreezedItems = (itemsToSort) => {
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/store/modules/shoots/getters.js:227]: I renamed the function to `orderFreezedItemsBySortedUIDsAtFreeze` because this is what the function actually does.
Moreover, I moved the logic that picks the item from the store <> freezedShootSkeletons to the store.

File: frontend/src/store/modules/shoots/index.js (hunk 5 above)
Comment (grolu) [reply frontend/src/store/modules/shoots/index.js:374]: We only want to copy values required to display the stale items in the table. We could also copy more data or even the whole objects. We should discuss this.

File: frontend/src/store/modules/shoots/getters.js [hunk 9]
```diff
@@ -257,11 +297,44 @@ export default {
...
+      })
+      sortByArrAtFreeze = sortByArr
+      sortDescArrAtFreeze = sortDescArr
+
+      return itemsToSort
+    }
+
+    return (items, sortByArr, sortDescArr) => {
+      // If the list is freezed, do not alter the order of the items, except if the user explicitly changes the sorting or when the filter changes
+      if (state.freezeSorting) {
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/store/modules/shoots/getters.js:315]: changed as described for better readability

File: frontend/src/store/modules/shoots/getters.js [hunk 10]
```diff
@@ -131,6 +139,18 @@ export function getSortVal (rootGetters, item, sortBy) {
 
 export default {
   filteredItems (state) {
+    if (state.freezeSorting) {
+      // When state is freezed, do not include new items
+      return compact(map(state.freezedShootSkeletons, skeletonItem => {
+        const storeItem = find(state.filteredShoots, item => {
+          return item.metadata.uid === skeletonItem.metadata.uid
+        })
```
Comment (petersutter) [inline frontend/src/store/modules/shoots/getters.js:147]:
```suggestion
        const storeItem = find(state.filteredShoots, ['metadata.uid', skeletonItem.metadata.uid])
```

File: frontend/src/store/modules/shoots/getters.js [hunk 11]
```diff
@@ -260,8 +301,45 @@ export default {
...
+      const searchStringChanged = items.length > sortedUIDsAtFreeze?.length
+
+      // If freezed, do not alter the order of the items
+      // except if the user explicitly changes the sorting or when the filter changes
+      if (!isEqual(sortByArrAtFreeze, sortByArr) || !isEqual(sortDescArrAtFreeze, sortDescArr) || searchStringChanged) {
+        // If the sorting or search has changed, sortedUIDsAtFreeze needs to be re-sorted according to the items order
+        const sortedItems = sortItems(items, sortByArr, sortDescArr)
+        sortedUIDsAtFreeze = map(sortedItems, ({ metadata }) => {
+          return metadata.uid
+        })
```
Comment (petersutter) [inline frontend/src/store/modules/shoots/getters.js:328]:
```suggestion
        sortedUIDsAtFreeze = map(sortedItems, 'metadata.uid')
```

File: frontend/src/store/modules/shoots/getters.js [hunk 12]
```diff
@@ -206,7 +229,25 @@ export default {
...
-    return (items, sortByArr, sortDescArr) => {
+    let sortedUIDsAtFreeze
+    let sortByArrAtFreeze
+    let sortDescArrAtFreeze
+
+    const orderFreezedItemsBySortedUIDsAtFreeze = (items) => {
+      return compact(map(sortedUIDsAtFreeze, freezedUID => {
+        const tableContainsItem = some(items, item => { // check that item shall still be visible in table
+          return item.metadata.uid === freezedUID
+        })
```
Comment (petersutter) [inline frontend/src/store/modules/shoots/getters.js:240]:
```suggestion
        const tableContainsItem = some(items, ['metadata.uid', freezedUID]) // check that item shall still be visible in table
```

File: frontend/src/store/modules/shoots/getters.js [hunk 13]
```diff
@@ -206,7 +229,25 @@ export default {
...
+        if (!tableContainsItem) {
+          return undefined // item not in table (filtered via table search)
+        }
+        return find(state.filteredShoots, item => {
+          return item.metadata.uid === freezedUID
+        })
```
Comment (petersutter) [inline frontend/src/store/modules/shoots/getters.js:246]:
```suggestion
        return find(state.filteredShoots, ['metadata.uid', freezedUID])
```

File: frontend/src/store/modules/shoots/getters.js (hunk 10 above)
Comment (grolu) [reply frontend/src/store/modules/shoots/getters.js:147]: Yes you are right we can simplify this for all the mentioned find functions. This changed a lot over time thus I did not yet simplify it. Will do so now...

File: frontend/src/store/modules/shoots/index.js [hunk 14]
```diff
@@ -459,6 +454,26 @@ const mutations = {
...
+
+    if (state.freezeSorting) {
+      const oldKeys = Object.keys(state.shoots)
+      const newKeys = Object.keys(shoots)
+      const removedShootKeys = difference(oldKeys, newKeys)
+      const addedShootKeys = difference(newKeys, oldKeys)
+
+      removedShootKeys.forEach(removedShootKey => {
+        const removedShoot = state.shoots[removedShootKey]
+        if (some(state.uidsAtFreeze.includes(removedShoot.metadata.uid))) {
```
Comment (petersutter) [inline frontend/src/store/modules/shoots/index.js:466]:
```suggestion
        if (state.uidsAtFreeze.includes(removedShoot.metadata.uid)) {
```

File: frontend/src/components/ShootListRow.vue [hunk 15]
```diff
@@ -265,4 +301,30 @@ export default {
...
+    left: 0;
+    position: absolute;
+    pointer-events: none;
+  }
+
+  .theme--light .stale .stale-overlay {
+    background-color: rgba(255,255,255,0.8)
+  }
+  .theme--dark .stale .stale-overlay {
+    background-color: rgba(0,0,0,0.8)
```
Comment (petersutter) [inline frontend/src/components/ShootListRow.vue:322]: Screenshot 2022-12-05 at 13 58 39
in dark mode it's a bit too dark

File: frontend/tests/unit/store.shoots.spec.js [hunk 16]
```diff
@@ -118,6 +129,12 @@ describe('store.shoots.getters', () => {
         }
       }
     ]
+
+    state = { shoots: fromPairs(map(shootItems, shootItem => [keyForShoot(shootItem.metadata), shootItem])), filteredShoots: shootItems, freezeSorting: false, freezedStaleShoots: [] }
```
Comment (petersutter) [inline frontend/tests/unit/store.shoots.spec.js:133]:
```suggestion
    state = {
      shoots: fromPairs(map(shootItems, shootItem => [keyForShoot(shootItem.metadata), shootItem])),
      filteredShoots: shootItems,
      freezeSorting: false,
      freezedStaleShoots: []
    }
```
added newlines for better readability

File: frontend/src/store/modules/shoots/getters.js [hunk 17]
```diff
@@ -131,6 +138,16 @@ export function getSortVal (rootGetters, item, sortBy) {
...
 export default {
   filteredItems (state) {
+    if (state.freezeSorting) {
+      // When state is freezed, do not include new items
+      return compact(map(state.uidsAtFreeze, freezedUID => {
+        const storeItem = find(state.filteredShoots, ['metadata.uid', freezedUID])
+        if (storeItem) {
+          return storeItem
+        }
+        return state.freezedStaleShoots[freezedUID]
```
Comment (petersutter) [inline frontend/src/store/modules/shoots/getters.js:148]: in my local test I had a case where no item was found in `state.freezedStaleShoots` for `freezedUID`

File: frontend/src/components/ShootListRow.vue (hunk 15 above)
Comment (grolu) [reply frontend/src/components/ShootListRow.vue:322]: Screenshot 2022-12-07 at 15 30 36

better?

Comment (petersutter) [reply frontend/src/components/ShootListRow.vue]: yes 👍

File: frontend/src/store/modules/shoots/getters.js [hunk 18]
```diff
@@ -144,6 +165,9 @@ export default {
   getShootListFilters (state) {
     return state.shootListFilters
   },
+  getFreezeSorting (state) {
```
Comment (holgerkoser) [inline frontend/src/store/modules/shoots/getters.js:168]: Why do we need this?

File: frontend/src/views/ShootList.vue [hunk 19]
```diff
@@ -165,6 +213,9 @@ export default {
     ...mapActions([
       'subscribeShoots'
     ]),
+    ...mapActions({
+      setFreezeSorting: 'shoots/setFreezeSorting'
+    }),
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:218]:
```suggestion
    ...mapActions('shoots', [
       'setFreezeSorting'
    ]),
```

File: frontend/src/views/ShootList.vue [hunk 20]
```diff
@@ -265,11 +325,14 @@ export default {
     ]),
     ...mapGetters('shoots', [
       'sortItems',
-      'searchItems'
+      'searchItems',
+      'getFreezeSorting',
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:329]:
```suggestion
```

File: frontend/src/views/ShootList.vue [hunk 21]
```diff
@@ -288,6 +351,14 @@ export default {
         }
       }
     },
+    freezeSorting: {
+      get () {
+        return this.getFreezeSorting
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:356]:
```suggestion
        return this.freezeSorting
```

File: frontend/src/views/ShootList.vue [hunk 22]
```diff
@@ -535,10 +629,18 @@ export default {
     items () {
       return this.cachedItems || this.mappedItems
     },
-    filtersDisabled () {
+    disableChangeFilters () {
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:632]: rename to `changeFiltersDisabled`

File: frontend/src/views/ShootList.vue [hunk 23]
```diff
@@ -288,6 +351,14 @@ export default {
         }
       }
     },
+    freezeSorting: {
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:354]: rename to `freezeSortingInternal`

File: frontend/src/views/ShootList.vue [hunk 24]
```diff
@@ -155,6 +194,15 @@ export default {
         this.$localStorage.removeItem(`project/${this.projectName}/shoot-list/options`) // clear project specific options
         this.$localStorage.setObject('projects/shoot-list/options', { sortBy, sortDesc, itemsPerPage })
       }
+    },
+    freezeSorting (value) {
+      if (value) {
+        const { sortBy, sortDesc } = this.options
+        const sortedItems = this.sortItems(this.items, sortBy, sortDesc)
+        this.sortedUIDsAtFreeze = map(sortedItems, 'metadata.uid')
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:202]: How is this related to the uids in the store? Why do we store the list of uids in two different locations?

File: frontend/src/views/ShootList.vue [hunk 25]
```diff
@@ -460,8 +543,18 @@ export default {
       })
     },
     allHeaders () {
-      const allHeaders = [...this.standardHeaders, ...this.customHeaders]
-      return sortBy(allHeaders, ['weight', 'text'])
+      let allHeaders = [...this.standardHeaders, ...this.customHeaders]
+      allHeaders = sortBy(allHeaders, ['weight', 'text'])
+      if (this.freezeSorting) {
+        return map(allHeaders, header => {
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:549]: I think this magic is done because otherwise we loose reactivity. We are mixing header configuration with reactive table header properties. I have the impression that in two weeks nobody will understand why do we create the deep copy here.

File: frontend/src/store/modules/shoots/index.js [hunk 26]
```diff
@@ -364,6 +367,17 @@ const actions = {
 
     commit('RESET_NEW_SHOOT_RESOURCE', shootResource)
     return state.newShootResource
+  },
+  setFreezeSorting ({ commit }, value) {
```
Comment (holgerkoser) [inline frontend/src/store/modules/shoots/index.js:371]: The store knows nothing about sorting.

File: frontend/src/views/ShootList.vue [hunk 27]
```diff
@@ -535,10 +629,18 @@ export default {
     items () {
       return this.cachedItems || this.mappedItems
     },
-    filtersDisabled () {
+    disableChangeFilters () {
+      return this.freezeSorting
+    },
+    hideFiltersForOnlyShootsWithIssues () {
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:635]: rename to showAllShoots

File: frontend/src/views/ShootList.vue [hunk 28]
```diff
@@ -535,10 +629,18 @@ export default {
...
       return !this.showOnlyShootsWithIssues
     },
-
+    filterTooltip () {
+      if (this.freezeSorting) {
+        return 'Filters cannot be changed when focus mode is active'
+      }
+      return undefined
```
Comment (holgerkoser) [inline frontend/src/views/ShootList.vue:642]:
```suggestion
      return this.freezeSorting
        ? 'Filters cannot be changed when focus mode is active'
         : ''
```

File: frontend/src/store/modules/shoots/getters.js (hunk 18 above)
Comment (grolu) [reply frontend/src/store/modules/shoots/getters.js:168]: not required anymore...

File: frontend/src/views/ShootList.vue (hunk 24 above)
Comment (grolu) [reply frontend/src/views/ShootList.vue:202]: This is due to historic reasons when we had the sorting in the getter.
Right now with the current changes there are two lists of UIDs. One unsorted in the store that is set when the user enabled the focus mode. And one in the component to cache the last sorting order. Let's discuss if we want to move the sorted UIDs to the store, or do something else there are multiple options.

**Review by petersutter — APPROVED** [review]
/lgtm

**Review by petersutter — APPROVED** [review]
/lgtm
